### PR TITLE
Add terminal block hash override

### PR DESCRIPTION
**Pull request by twoeths**
**Motivation**

Implement this merge spec https://github.com/ethereum/consensus-specs/pull/2617/files

**Description**

part of  #3076

### CODE REVIEW

File: packages/lodestar/test/unit/eth1/eth1MergeBlockTracker.test.ts [hunk 1]
```diff
@@ -13,19 +15,79 @@ describe("eth1 / Eth1MergeBlockTracker", () => {
...
-    SECONDS_PER_SLOT: 0,
-    // Hardcode TTD to a low value
-    TERMINAL_TOTAL_DIFFICULTY: BigInt(terminalTotalDifficulty),
-  } as Partial<IChainConfig>) as IChainConfig;
-
+  let config: IChainConfig;
   let controller: AbortController;
-  beforeEach(() => (controller = new AbortController()));
+  beforeEach(() => {
+    controller = new AbortController();
```
Comment (dapplion) [inline packages/lodestar/test/unit/eth1/eth1MergeBlockTracker.test.ts]: Keep this before each as before as the block 
```ts
  let controller: AbortController;
  beforeEach(() => (controller = new AbortController()));
  afterEach(() => controller.abort());
```
And add new code in another beforeEach block below

File: packages/lodestar/test/e2e/eth1/eth1MergeBlockTracker.test.ts [hunk 2]
```diff
@@ -9,6 +9,7 @@ import {testnet} from "../../utils/testnet";
 import {testLogger} from "../../utils/logger";
 import {quantityToBigint} from "../../../src/eth1/provider/utils";
 import {ZERO_HASH} from "../../../src/constants";
+import {fromHexString} from "@chainsafe/ssz";
```
Comment (dapplion) [inline packages/lodestar/test/e2e/eth1/eth1MergeBlockTracker.test.ts]: Sort imports from dependencies above file imports

File: packages/lodestar/src/eth1/eth1MergeBlockTracker.ts [hunk 3]
```diff
@@ -172,6 +176,13 @@ export class Eth1MergeBlockTracker {
   }
 
   private async fetchPreviousBlocks(): Promise<void> {
+    // Terminal block hash override takes precedence over terminal total difficulty
+    if (!ssz.Root.equals(this.config.TERMINAL_BLOCK_HASH, ZERO_HASH)) {
+      const powBlockOverride = await this.getPowBlock(toHexString(this.config.TERMINAL_BLOCK_HASH));
+      if (powBlockOverride) {
+        return this.setTerminalPowBlock(powBlockOverride);
+      }
+    }
```
Comment (dapplion) [inline packages/lodestar/src/eth1/eth1MergeBlockTracker.ts]: I would do this in the constructor, and then return such that no more work is done in this module

**Comment by dapplion** [comment]
@tuyennhv Setting TERMINAL_BLOCK_HASH means that no matter what that's the merge block hash.

So, change the stored merge block, to merge block hash. Then in the constructor just do

```ts
constructor() {
  if (TERMINAL_BLOCK_HASH !== 0) {
    setMergeBlockHash(TERMINAL_BLOCK_HASH)
    return
  }
}
```

File: packages/lodestar/src/eth1/eth1MergeBlockTracker.ts [hunk 4]
```diff
@@ -142,8 +143,12 @@ export class Eth1MergeBlockTracker {
     this.close();
   }
 
-  private startFinding(): void {
+  private async startFinding(): Promise<void> {
     if (this.status !== StatusCode.PRE_MERGE) return;
+    const powBlockOverride = await this.getPowBlock(toHexString(this.config.TERMINAL_BLOCK_HASH));
```
Comment (dapplion) [inline packages/lodestar/src/eth1/eth1MergeBlockTracker.ts]: This getPowBlock is likely to throw and startFinding is not error handled. I would catch this getPowBlock() and if it throws, just continue and try to find the merge block

File: packages/lodestar/src/eth1/eth1MergeBlockTracker.ts [hunk 5]
```diff
@@ -142,8 +143,21 @@ export class Eth1MergeBlockTracker {
     this.close();
   }
 
-  private startFinding(): void {
+  private async startFinding(): Promise<void> {
     if (this.status !== StatusCode.PRE_MERGE) return;
+    try {
```
Comment (dapplion) [inline packages/lodestar/src/eth1/eth1MergeBlockTracker.ts]: Can you add space before and after this try {} catch {} block? Also a short explanatory comment of what's going on with `TERMINAL_BLOCK_HASH`

File: packages/lodestar/src/eth1/eth1MergeBlockTracker.ts [hunk 6]
```diff
@@ -215,7 +229,7 @@ export class Eth1MergeBlockTracker {
             // Is terminal total difficulty block
             if (childBlock.parentHash === parentBlock.blockhash) {
               // AND has verified block -> parent relationship
-              return this.setMergeBlock(childBlock);
+              return this.setTerminalPowBlock(childBlock);
```
Comment (dapplion) [inline packages/lodestar/src/eth1/eth1MergeBlockTracker.ts:238]: Just realized that if `TERMINAL_BLOCK_HASH` is set all this logic is not valid. Now there is a completely different condition to find the terminal POW block

Comment (twoeths) [reply packages/lodestar/src/eth1/eth1MergeBlockTracker.ts:238]: if a `TERMINAL_BLOCK_HASH` is configured, this is not reached

**Review by dapplion — APPROVED** [review]

**Comment by twoeths** [comment]
I can't merge this due to `license/cla` check, @dapplion do you have any ideas?

**Pull request merged**